README.md

```diff
@@ -189,6 +189,149 @@ We only did a small refactor, so the build tab running `npm run build` should
+
+Ensure your application builds successfully (`Success` in the `npm run watch`
+terminal). We may need to adjust these types later when we decode the data.
+
+## Step 4: Decoding data (the hard way)
+
+Once you have these types defined, we will be adding some code to take the
+values from the JSON and decode them into values of the types we have defined.
+
+First, install the
+[melange-json](https://github.com/melange-community/melange-json) library:
+
+```sh
+opam install melange-json
+```
+
+Then, modify the `libraries` field in the `src/dune` file to include
+`melange-json`:
+
+```dune
+(libraries melange-json reason-react)
+```
+
+Now we’re ready to decode JSON into our types. Add a `Decode` module inside
+`Feed.re` with the following code:
+
+```reason
+module Decode = {
+  let link = json =>
+    Json.Decode.{
+      Link.href: json |> field("href", string),
+      title: json |> field("title", string),
+    };
+  let entry = json =>
+    Json.Decode.{
+      Entry.content: json |> optional(field("content", string)),
+      links: json |> field("links", list(link)),
+      title: json |> field("title", string),
+      updated: json |> field("updated", float),
+    };
+
+  let feed = json =>
+    Json.Decode.{entries: json |> field("entries", list(entry))};
+};
+
+let data = {| {
+  "entries": [
+    {
+      "content": "<div>Hello</div>",
+      "links": [
+        {
+          "title": "",
+          "href": "https://github.com/melange-community/melange-json"
+        }
+      ],
+      "title": "jchavarri starred melange-community/melange-json",
+      "updated": 1723639727000.0
+    }
+  ]
+} |};
+
+let demoFeed =
+  try(Some(data |> Json.parseOrRaise |> Decode.feed)) {
+  | Json.Decode.DecodeError(msg) =>
+    Js.log(msg);
+    None;
+  };
+```
+
+This code demonstrates several key features:
+- The [pipe
+  operator](https://melange.re/v4.0.0/communicate-with-javascript.html#pipe-last)
+  `|>` sends the value on its right as the last argument to the function on the
+  left, enabling clean, step-by-step data processing.
+- Local module [opens](https://reasonml.github.io/docs/en/module#local-opens)
+  like `Json.Decode.{ ... }` let us simplify the code by avoiding repetitive
+  module prefixes.
+- Exception handling uses `try` similar to JavaScript, but with [pattern
+  matching](https://reasonml.github.io/docs/en/pattern-matching) for the `catch`
+  part.
+- To ensure consistent return types in the exception handling, we leverage
+  [optional types](https://reasonml.github.io/docs/en/option).
+
+### Step 4 completion check
+
+Add this debug line at the bottom of `App.re`:
+
+```reason
+Js.log(Feed.demoFeed)
+```
+
+When you open the browser to view the page, check the console for the logged
+content. You should see something like this:
+
+![Console log - printing some runtime values](README-imgs/console-log.png)
+
+> [!NOTE]
+> 
+> Can you figure out why the `hd` and `tl` values appear in the `entries` record
```

Comment:
I think the reason for using a list over an array will not be very clear at this point. Maybe just use array for now, and then add switching to list as an exercise?

Reply:
https://github.com/ahrefs/react-alicante-workshop/pull/3/commits/49637671ffc4cca25e31186069f0c6bb55a1ab9b